### ISSUE TITLE
Bumping 2.6.3 to 2.6.4 (kamon-bundle and kamon-prometheus) breaks logging in our apps (Akka HTTP, Play)

### ISSUE THREAD

**Issue by przemek-pokrywka** [kamon-io/Kamon#1297]
Hello,

Bumping Kamon bundle and prometheus to 2.6.4 (from 2.6.3) breaks logging in our applications.
The Akka-HTTP-based app (Akka HTTP 10.2.10) simply does not log anything.
The Play-based app (Play 2.8.20) crashes during initialization, with the following stack trace:

```
Oops, cannot start the server.
java.lang.ClassCastException: class org.slf4j.helpers.NOPLoggerFactory cannot be cast to class ch.qos.logback.classic.LoggerContext (org.slf4j.helpers.NOPLoggerFactory and ch.qos.logback.classic.LoggerContext are in unnamed module of loader 'app')
	at play.api.libs.logback.LogbackLoggerConfigurator.configure(LogbackLoggerConfigurator.scala:91)
	at play.api.libs.logback.LogbackLoggerConfigurator.configure(LogbackLoggerConfigurator.scala:73)
	at play.api.inject.guice.GuiceApplicationBuilder.$anonfun$configureLoggerFactory$1(GuiceApplicationBuilder.scala:137)
	at scala.Option.map(Option.scala:242)
	at play.api.inject.guice.GuiceApplicationBuilder.configureLoggerFactory(GuiceApplicationBuilder.scala:136)
	at play.api.inject.guice.GuiceApplicationBuilder.applicationModule(GuiceApplicationBuilder.scala:112)
	at play.api.inject.guice.GuiceBuilder.injector(GuiceInjectorBuilder.scala:201)
	at play.api.inject.guice.GuiceApplicationBuilder.build(GuiceApplicationBuilder.scala:155)
	at play.api.inject.guice.GuiceApplicationLoader.load(GuiceApplicationLoader.scala:21)
	at play.core.server.ProdServerStart$.start(ProdServerStart.scala:53)
	at play.core.server.ProdServerStart$.main(ProdServerStart.scala:29)
	at play.core.server.ProdServerStart.main(ProdServerStart.scala)
```

Here is the problematic version bump - when we revert it, the problems disappear.
```diff
index 4eba767..98b688f 100644
--- a/project/Common.scala
+++ b/project/Common.scala
@@ -91,8 +91,8 @@ object Common {
       project
         .settings(
           libraryDependencies ++= Seq(
-            "io.kamon" %% "kamon-bundle"     % "2.6.3",
-            "io.kamon" %% "kamon-prometheus" % "2.6.3"
+            "io.kamon" %% "kamon-bundle"     % "2.6.4",
+            "io.kamon" %% "kamon-prometheus" % "2.6.4"
           )
         )
   }
```

Comparison of the dependency trees brings no insights, all that differs is the Kamon artifact versions, that's it.
The logging libraries used by the Play app:
```
ch.qos.logback.logback-classic-1.4.8.jar
ch.qos.logback.logback-core-1.4.8.jar
dev.zio.zio-logging-slf4j_2.13-2.1.14.jar
dev.zio.zio-logging_2.13-2.1.14.jar
org.slf4j.jcl-over-slf4j-1.7.36.jar
org.slf4j.jul-to-slf4j-1.7.36.jar
org.slf4j.log4j-over-slf4j-2.0.7.jar
org.slf4j.slf4j-api-2.0.7.jar
```
Scala version: 2.13.12

**Comment by Dichotomia** [comment]
Exact same problem here

**Comment by irajhedayati** [comment]
Same here. I'm using MDC, and my context variables always return `null`. Looking at the code, there are some dependency issues. The log4j classes are present in the `kammon-bundle`. I even tried to exclude the library but didn't work.

<img width="1020" alt="image" src="https://github.com/kamon-io/Kamon/assets/10663735/cdfc53f2-5fb1-4d6b-b4e2-b71615f626f6">

**Comment by hughsimpson** [comment]
Welp this is unexpected. Let me have a look, and apologies - I was responsible for that release 😅

**Comment by hughsimpson** [comment]
Nothing jumps out... I thought maybe there's a stray akka reference in the pekko branch, but a great reveals nothing but some copied comments that haven't been suitably adapted...

**Comment by hughsimpson** [comment]
Well the only thing that changed that I could think of as relevant is the sbt version. It's possible that some semantic correction has led to a 'provided' dependency actually.. being... provided. Suggest checking the dependencies of the modules you use, and explicitly including what ever logback libs you were implicitly depending on before...

**Comment by hughsimpson** [comment]
A microproject that emulated this problem would be immensely helpful anyway. Since there are multiple thumbs up on this, I assume that someone should be able to whip one up. I'll dig in as soon as I can.

**Comment by irajhedayati** [comment]
@hughsimpson, thanks for looking into this. Here is my issue:

**build.properties**
```scala
sbt.version = 1.6.2
```

**build.sbt**

```scala
ThisBuild / version := "0.1.0-SNAPSHOT"

ThisBuild / scalaVersion := "2.13.12"

val LogbackVersion = "1.4.11"
val KamonVersion = "2.6.4"
val ScalaLoggingVersion = "3.9.5"

lazy val root = (project in file("."))
  .settings(
    name := "untitled",
    libraryDependencies ++= Seq(
      "com.typesafe.scala-logging" %% "scala-logging"    % ScalaLoggingVersion,
      "ch.qos.logback"              % "logback-classic"  % LogbackVersion,
      "io.kamon"                   %% "kamon-bundle"     % KamonVersion,
      "io.kamon"                   %% "kamon-prometheus" % KamonVersion
    )
  )
```

**src/main/resources/logback.xml**
```xml
<configuration>
    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
        <encoder>
            <pattern>%d{yyyy-MM-dd HH:mm:ss.SSS} %-5level%X{runId}%X{requestId} %-36.36logger{36}  %msg%n%rEx</pattern>
        </encoder>
    </appender>

    <root level="INFO">
        <appender-ref ref="STDOUT"/>
    </root>
</configuration>
```

**src/main/scala/Main.scala**
```scala
import com.typesafe.scalalogging.LazyLogging
import org.slf4j.MDC

import java.util.UUID

object Main extends App with LazyLogging {
  MDC.put("runId", s" [runId=${UUID.randomUUID().toString}]")
  MDC.put("requestId", s" [requestId=${UUID.randomUUID().toString}]")
  logger.info("Sample message")
  MDC.clear()

}
```

## Kamon 2.6.3

Here is the output

```
2023-10-08 23:02:19.661 INFO  [runId=2c7cc229-afb2-4d8e-9a18-c34eda1af7e9] [requestId=7434964a-e58d-40ce-aae8-d46bb2b13335] Main$                                 Sample message

Process finished with exit code 0
```

## Kamon 2.6.4

Here is the output

```
SLF4J: Failed to load class "org.slf4j.impl.StaticMDCBinder".
SLF4J: Defaulting to no-operation MDCAdapter implementation.
SLF4J: See http://www.slf4j.org/codes.html#no_static_mdc_binder for further details.
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.

Process finished with exit code 0
```

------

I also checked the classpath, and everything is the same. I think your suspicion about some classes creeping into the Kamon is the cause.

Let me know if you want me to upload the project somewhere, but it's basically just these files.

**Comment by hughsimpson** [comment]
Thanks this was very helpful. I believe I've identified the issue and am working on a patched release

**Comment by hughsimpson** [comment]
Marking this as resolved now - testing with 2.6.5 shows this to be fixed
